Add downgrade migration without superuser privileges

piercefreeman opened the pull request:
As mentioned in the docs:

> Generated migrations only work in one direction, since without a superuser account and a bunch of dangerous operations it isn't possible to remove PostgreSQL ENUM values.

One workaround to the above is to archive the current enum definition and re-create it as a fresh value. I made some minor changes to the `SyncEnumValuesOp` to support this behavior, namely:

- Keeping track of the tables that reference each enum, since a per-table migration is required to switch it to the new enum field
- Track state of forward/reverse migrations so we can perform the correct logic. You technically could switch to using this rename+create logic for both forward and reverse passes, but I prefer the explicitness of the `ALTER TYPE ... ADD VALUE` for the forward migrations since this is what I do when I'm doing manual SQL migrations.

I've also added some typedefinitions and dataclasses to make passing data dictionaries a bit easier to debug. This also necessarily means that these changes only support recent Python versions. I can either strip these out for the sake of gaining downgrades without constraining Python versions - or this release can be semversioned to indicate its incompatibility with previous versions. Let me know which course you'd rather take @dw.

dw commented:
This looks lovely. Will review when I get a sec, feel free to chase after a week

Aside from Python compatibility, how does this work for users who already have existing migrations? Will bumping the library version from an older release still work okay

piercefreeman commented:
Good point. The main thing that would break for old migrations was the `sync_enum_values` contract since we added new params. I've added default values for these in [3421058](https://github.com/dw/alembic-autogenerate-enums/pull/7/commits/3421058a02a69fa6a94d2cdad9f8340312d50125) that should maintain the old behavior unless we have a new autogenerated version of the sync_enum_values operation call.

dw commented:
Just realized I don't have an active project to test this with. The code changes look fine from eye inspection, happy with Python version bump since it's easy to remove if anyone complains. Happy to merge if you're happy with it

piercefreeman commented:
@dw I put together a test harness to make sure that these migrations are working as advertised. You can see the results of the latest commit here:
https://github.com/piercefreeman/alembic-autogenerate-enums/actions/runs/4569712098

The included migration tests for:
- The old API contract of forward migrations
- The new API contract of forward migrations
- The new API contract of reverse migrations

The github action runner is set up to run on every commit, so these should get auto-launched when there are additional commits / PRs going forward. Also bumped the version to 0.1.0.

Ready to merge in and release when you are.

piercefreeman commented on test-harness/alembic/versions/2635ee8b59d5_deprecated_upgrade.py:
```diff
@@ -0,0 +1,29 @@
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    #op.sync_enum_values('public', 'simpleenum', ['A', 'B', 'C', 'D'], ['A', 'B', 'C'])
```
@dw Excluding the v0.0.2 downgrade test since we didn't expect it to work under most postgres conditions.

dw commented:
Amazing work! Thank you

piercefreeman commented:
Thanks :) You might have to opt-in on your end for the github actions to start running, but I'll bow out for now.

dw commented:
LMK if you'd like added as outside contributor, I know plenty of folk are using this code but I haven't touched it in years.

piercefreeman commented:
Sure! Happy to groom tickets and help out where I can.
